pppd/eap.c

```diff
@@ -2304,6 +2521,103 @@ int len;
 			TIMEOUT(eap_rechallenge, esp, esp->es_rechallenge);
 		break;
 
+#ifdef CHAPMS
+	case EAPT_MSCHAPV2:
+		if (len < 1) {
+			error("EAP: received MSCHAPv2 with no data");
+			eap_figure_next_state(esp, 1);
+			break;
+		}
+		GETCHAR(opcode, inp);
+		len--;
+
+		switch (opcode) {
+		case CHAP_RESPONSE:
+			if (esp->es_server.ea_state != eapMSCHAPv2Chall) {
+				error("EAP: unexpected MSCHAPv2-Response");
+				eap_figure_next_state(esp, 1);
+				break;
+			}
+			/* skip MS ID + len */
+			INCPTR(3, inp);
+			GETCHAR(vallen, inp);
+			len -= 4;
+
+			if (vallen != MS_CHAP2_RESPONSE_LEN) {
+				error("EAP: Invalid MSCHAPv2-Response "
+						"length %d", vallen);
+				eap_figure_next_state(esp, 1);
+				break;
+			}
+
+			/* Not so likely to happen. */
+			if (vallen >= len + sizeof (rhostname)) {
```

Comment:
This is a repeat of the buffer overflow bug that resulted in a CVE, that was fixed in commit 8d7970b8f3db.

Reply:
Thanks for the review @paulusmack 

I updated the commit with the following patch:

```diff
@@ -2541,7 +2527,7 @@
                        GETCHAR(vallen, inp);
                        len -= 4;

-                       if (vallen != MS_CHAP2_RESPONSE_LEN) {
+                       if (vallen != MS_CHAP2_RESPONSE_LEN || vallen > len) {
                                error("EAP: Invalid MSCHAPv2-Response "
                                                "length %d", vallen);
                                eap_figure_next_state(esp, 1);
@@ -2549,7 +2535,7 @@
                        }

                        /* Not so likely to happen. */
-                       if (vallen >= len + sizeof (rhostname)) {
+                       if (len - vallen >= sizeof (rhostname)) {
                                dbglog("EAP: trimming really long peer name down");
                                BCOPY(inp + vallen, rhostname, sizeof (rhostname) - 1);
                                rhostname[sizeof (rhostname) - 1] = '\0';
```

I also converted function definitions to ANSI C style.